src/nb_type.cpp

```diff
@@ -342,11 +368,19 @@ type_data *nb_type_c2p(nb_internals *internals_,
 }
 
 void nb_type_unregister(type_data *t) noexcept {
-    nb_type_map_slow &type_c2p_slow = internals->type_c2p_slow;
-    nb_type_map_fast &type_c2p_fast = internals->type_c2p_fast;
+    nb_internals *internals_ = internals;
+    nb_type_map_slow &type_c2p_slow = internals_->type_c2p_slow;
 
-    size_t n_del_slow = type_c2p_slow.erase(t->type),
-           n_del_fast = type_c2p_fast.erase(t->type);
+    lock_internals guard(internals_);
+    size_t n_del_slow = type_c2p_slow.erase(t->type);
+
+#if defined(NB_FREE_THREADED)
+    // In free-threaded mode, stale type information remains in the
+    // 'type_c2p_fast' TLS. It is freed when the thread terminates.
```

Comment:
The problem is that one thread can unregister a type, but many other threads could have that type in their caches. So you would need to delete it there as well, which requires synchronization. However, the point of this construction is to avoid synchronization.

The comment here is missing an important detail: in free-threaded mode, we immortalize type objects in `type_new()` and let them leak.  This means that this function will only ever be called to delete the `type_data*` of C++ types that are subclassed in Python, and these don't have a unique entry in `type_c2p_fast`.

Reply:
A related question would be what to do if we don't immortalize type objects at some point in the future (which is unfortunate because of the leaks). Is there a way of having fast unlooked lookups while catching the case you mention.

The only think I can think of is that there is some kind of global version counter incremented in `nb_type_unregister`, which is also cached in the TLS. If the version is different, the cache must be cleared. This would work most of the time but is still prone to a race when one thread is unregistering a type, while the other one is currently performing a lookup in the cache.

Reply:
I agree this is safe now because we immortalize. It would be nice to mention that in the comment here.

As far as the potentially non-immortal future goes: one useful trick here is to increment the counter once before you remove the type from the c2p_slow map, and once again afterward. If lookup finds the counter at an odd value, it must retry. I think this is safe as long as you use an atomic value for the counter, guard unregistration with the internals mutex so multiple threads can't unregister simultaneously, and use acquire ordering for the first increment and release for the second.